document/js-api/index.bs

```diff
@@ -1149,27 +1143,42 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
             1. Return [=ref.null=] |heaptype|.
         1. Otherwise,
             1. Throw a {{TypeError}}.
-    1. If |type| is a subtype of [=ref=] |null| [=heap-type/func=],
+    1. If |heaptype| matches [=heap-type/func=],
         1. If |v| is an [=Exported Function=],
             1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
             1. Let |testresult| be [=ref_test=](|heaptype|, [=ref.func=] |funcaddr|).
             1. If |testresult| is true,
                 1. Return [=ref.func=] |funcaddr|.
         1. Throw a {{TypeError}}.
-    1. If |heaptype| is [=extern=],
-        1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
-        1. If a [=extern address=] |externaddr| exists such that |map|[|externaddr|] is the same as |v|,
-            1. Return [=ref.extern=] |externaddr|.
-        1. Let [=extern address=] |externaddr| be the smallest address such that |map|[|externaddr|] [=map/exists=] is false.
-        1. [=map/Set=] |map|[|externaddr|] to |v|.
-        1. Return [=ref.extern=] |externaddr|.
-    1. If |type| is a subtype of [=ref=] |null| [=any=],
-        1. Let |externref| be [=ToWebAssemblyValue=](|v|, [=ref=] [=extern=]).
-        1. Let |internref| be [=extern_internalize=](|externref|).
-        1. Let |testresult| be [=ref_test=](|heaptype|, |internref|).
-        1. If |testresult| is true,
-            1. Return |internref|.
-        1. Throw a {{TypeError}}.
+    1. If |heaptype| matches [=any=],
+        1. If |v| [=is a Number=],
+            1. Let |i32| be ? [$ToInt32$](|v|).
+            1. If |v| is equal to |i32| and [=ℝ=](|v|) < 2<sup>30</sup> and [=ℝ=](|v|) ⩾ -2<sup>30</sup>,
+                1. If [=i31=] does not match |heaptype|,
+                    1. Throw a {{TypeError}}.
+                1. Return [=ref.i31=] |v|.
+        1. If |v| is an [=Exported GC Object=],
+            1. Let |objectaddr| be the value of |v|'s \[[ObjectAddress]] internal slot.
+            1. Let |objecttype| be the value of |v|'s \[[ObjectType]] internal slot.
+            1. If |objecttype| is "array",
+                1. If [=array=] does not match |heaptype|,
+                    1. Throw a {{TypeError}}.
+                1. Return [=ref.array=] |objectaddr|.
```

Comment:
Don't we still need to do a ref.test to make sure the downcast succeeds when |heaptype| is a specific array type? Same with structs below.

Reply:
Well, we do not support concrete array types in the API yet, so this assumes the invariant that heaptype is one of the abstract heap types. Should we make this more powerful?

Reply:
Hmm, am I misremembering the result of those discussions? I thought we allowed arbitrary types on the boundary.

Reply:
Yes, I think I was correct: https://github.com/WebAssembly/gc/issues/279#issuecomment-1349574128

Reply:
Or maybe I misremember -- I assumed concrete types cannot come up because we cannot express them API-side. But did we decide to allow calling into exported Wasm functions with concrete types?

I guess there is nothing else preventing it anyway, so something needs to be fixed here. I'll extend the semantics. However, this will probably require replacing the existing [[ObjectKind]] thingy with a proper call to the embedder interface (to get the precise object type), which I haven't written yet.

Reply:
Ah, right, we don't have a way to describe concrete types from the JS API, but yes, I was thinking about the case of calling into an exported function (or setting an exported global, etc) that expects a concrete type.

Reply:
Ah yes, you are right, I already added that as part of the funcref proposal.

Reply:
Hm, actually I didn't. So that probably is a dangling ref at the moment.

Reply:
Okay, I think this is fixed now. I also refactored the function to have less duplication.